core/node/node_framework/src/service/mod.rs

```diff
@@ -114,55 +124,121 @@ impl ZkStackService {
             let name = task.name().to_string();
-            let after_node_shutdown = task.after_node_shutdown();
-            let task_future = Box::pin(task.run(self.stop_receiver()));
-            let task_repr = TaskRepr {
-                name,
-                task: Some(task_future),
-                after_node_shutdown,
-            };
-            tasks.push(task_repr);
+            let stop_receiver = self.stop_receiver();
+            let task_barrier = task_barrier.clone();
+            let task_future = Box::pin(async move {
+                task.run_with_barrier(stop_receiver, task_barrier)
+                    .await
+                    .with_context(|| format!("Task {name} failed"))
+            });
+            tasks.push(task_future);
         }
         if tasks.is_empty() {
             anyhow::bail!("No tasks to run");
         }
+        // Finally, the preconditions are just futures.
+        for precondition in std::mem::take(&mut self.preconditions) {
+            let name = precondition.name().to_string();
+            let stop_receiver = self.stop_receiver();
+            let task_barrier = task_barrier.clone();
+            let task_future = Box::pin(async move {
+                precondition
+                    .check_with_barrier(stop_receiver, task_barrier)
+                    .await
+                    .with_context(|| format!("Precondition {name} failed"))
+            });
+            // Internally we use the same representation for tasks and preconditions.
+            preconditions.push(task_future);
+        }
 
         // Wiring is now complete.
         for resource in self.resources.values_mut() {
             resource.stored_resource_wired();
         }
+        tracing::info!("Wiring complete");
+
+        // Launch precondition checkers. Generally, if all of them succeed, we don't need to do anything.
+        // If any of them fails, the node have to shut down.
+        // So we run them in a detached task with an oneshot channel to report the error, if it occurs,
+        // and we'll use this channel in the same way that we use to check if any of the task exits.
+        // The difference is that preconditions, unlike tasks, are allowed to exit without causing the node
+        // to shut down.
+        let (precondition_fail_sender, precondition_fail_receiver) = oneshot::channel();
+        self.runtime.spawn(async move {
+            let preconditions = preconditions.into_iter().map(|fut| async move {
+                // Spawn each precondition as a separate task.
+                // This way we can handle the cases when a precondition task panics and propagate the message
+                // to the service.
+                // This is important since this task is detached: we do not store a handle for it and never await it explicitly.
+                let handle = tokio::runtime::Handle::current();
+                match handle.spawn(fut).await {
```

Comment:
On using a handle: over time, I kind of came to the conclusion that if spawning panics, it's a bit more obvious when it panics when you try obtain the handle, e.g. separation of concepts for getting the handle and actually spawning the future. It's a minor thing and mostly my preference, but can change if you think it's important.

On `catch_unwind` -- I've actually tried using it first, but unfortunately it requires an [UnwindSafe](https://doc.rust-lang.org/nightly/core/panic/unwind_safe/trait.UnwindSafe.html) bound on the futures. When I played with it, I met a few compiler errors when this condition was not met, and they were extremely unfriendly to non-async-native people, so adding an `UnwindSafe` bound to the precondition doesn't seem to be a good solution because of the complexity it bears.